# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


**Гіпотези:**   

H₀: μ = 123,7 (середній час від першого заходу на сайт до покупки (time-to-purchase) становив 123,7 хвилин)  
H₁: μ > 123,7 (середній час до покупки користувачів збільшився)

**Тип тесту:** правосторонній (one-tailed)  

Стверджується, що середній час від першого заходу на сайт до покупки (time-to-purchase) користувачів **збільшився**, а не просто відрізняється.

σ - невідома, вибірка мала (n=7), тому використовуємо **t-тест**:

**T = (x̄ - μ₀) / (s / √n)**

**t-статистика:**

μ₀ = 123,7 - середнє значення

n = 7 - розмір вибірки

α = 0,1 - рівень значущості

x̄ = 132,29 - вибіркове середнє значення

s = 10,98 - вибіркове стандартне відхилення

t = (x̄ - μ₀) / (s / √n) = (132,29 - 123,7) / (10,98 / √7) = 2,069 - t-статистичне (розрахункове)

df = 6 - ступені свободи

t' = 1,440 - t-критичне

p-value = 0,04202

In [1]:
import numpy as np
import pandas as pd
from scipy import stats

#Дані
customers = np.array([128, 135, 121, 142, 126, 151, 123])
mu0 = 123.7   #середній час
alpha = 0.10

#Вибіркова статистика
x_bar = np.mean(customers)
s = np.std(customers, ddof=1)   #стандартне відхилення з n-1
n = len(customers)

#t-статистика
t_stat = (x_bar - mu0) / (s / np.sqrt(n))

#Критичне значення для правостороннього тесту
df = n - 1
t_crit = stats.t.ppf(1-alpha, df)

#p-value
p_value = 1 - stats.t.cdf(t_stat, df)

#Виведення значень
print(f"Середнє значення: μ₀ = {mu0}")
print(f"Розмір вибірки: n = {n}")
print(f"Рівень значущості: α = {alpha}")
print(f"Вибіркове середнє: x̄ = {x_bar:.2f}")
print(f"Вибіркове стандартне відхилення: s = {s:.2f}")
print(f"t-статистика: t = {t_stat:.3f}")
print(f"Критичне значення (df={df}): t' = {t_crit:.3f}")
print(f"p-value: p-value = {p_value:.5f}")

#Висновок
if t_stat > t_crit:
    print(f"Відхиляємо H0: середній час від першого заходу на сайт до покупки (time-to-purchase) користувачів збільшився на рівні значущості {alpha}")
    print(f"Висновок: редизайн сайту збільшив середній час від першого заходу на сайт до покупки (time-to-purchase) для користувачів")
else:
    print("Не відхиляємо H0: доказів недостатньо")
    print(f"Висновок: редизайн сайту не збільшив середній час від першого заходу на сайт до покупки (time-to-purchase) для користувачів")

Середнє значення: μ₀ = 123.7
Розмір вибірки: n = 7
Рівень значущості: α = 0.1
Вибіркове середнє: x̄ = 132.29
Вибіркове стандартне відхилення: s = 10.98
t-статистика: t = 2.069
Критичне значення (df=6): t' = 1.440
p-value: p-value = 0.04202
Відхиляємо H0: середній час від першого заходу на сайт до покупки (time-to-purchase) користувачів збільшився на рівні значущості 0.1
Висновок: редизайн сайту збільшив середній час від першого заходу на сайт до покупки (time-to-purchase) для користувачів


**Гіпотези:**   

H₀: μ = 123,7 (середній час від першого заходу на сайт до покупки (time-to-purchase) становив 123,7 хвилин)  
H₁: μ > 123,7 (середній час до покупки користувачів збільшився)

**Тип тесту:** правосторонній (one-tailed)  

Стверджується, що середній час від першого заходу на сайт до покупки (time-to-purchase) користувачів **збільшився**, а не просто відрізняється.

σ - невідома, вибірка мала (n=7), тому використовуємо **t-тест**:

**T = (x̄ - μ₀) / (s / √n)**

**t-статистика:**

μ₀ = 123,7 - середнє значення

n = 7 - розмір вибірки

α = 0,05 - рівень значущості

x̄ = 132,29 - вибіркове середнє значення

s = 10,98 - вибіркове стандартне відхилення

t = (x̄ - μ₀) / (s / √n) = (132,29 - 123,7) / (10,98 / √7) = 2,069 - t-статистичне (розрахункове)

df = 6 - ступені свободи

t' = 1,943 - t-критичне

p-value = 0.04202

In [2]:
#Дані
customers_1 = np.array([128, 135, 121, 142, 126, 151, 123])
mu0_1 = 123.7   #середній час
alpha_1 = 0.05

#scipy тест
result = stats.ttest_1samp(customers_1, popmean=mu0_1, alternative="greater")
t_crit_1 = stats.t.ppf(1 - alpha_1, result.df)

#Виведення результатів
print(f"SCIPY:")
print(f"t-статистика: t = {result.statistic:.3f}")
print(f"Критичне значення (df={df}): t' = {t_crit_1:.3f}")
print(f"p-value: p-value = {result.pvalue:.5f}")

#Висновок на основі р-значення
print(f"\nВисновок на основі р-значення:")
if result.pvalue < alpha_1:
    print(f"Відхиляємо H0: середній час від першого заходу на сайт до покупки (time-to-purchase) користувачів збільшився на рівні значущості {alpha_1}")
    print(f"Висновок: редизайн сайту збільшив середній час від першого заходу на сайт до покупки (time-to-purchase) для користувачів")
else:
    print("Не відхиляємо H0: доказів недостатньо")
    print(f"Висновок: редизайн сайту не збільшив середній час від першого заходу на сайт до покупки (time-to-purchase) для користувачів")

#Висновок на основі t-статистичного та t-критичного
print(f"\nВисновок на основі t-статистичного та t-критичного:")
if result.statistic > t_crit_1:
    print(f"Відхиляємо H0: середній час від першого заходу на сайт до покупки (time-to-purchase) користувачів збільшився на рівні значущості {alpha_1}")
    print(f"Висновок: редизайн сайту збільшив середній час від першого заходу на сайт до покупки (time-to-purchase) для користувачів")
else:
    print("Не відхиляємо H0: доказів недостатньо")
    print(f"Висновок: редизайн сайту не збільшив середній час від першого заходу на сайт до покупки (time-to-purchase) для користувачів")

SCIPY:
t-статистика: t = 2.069
Критичне значення (df=6): t' = 1.943
p-value: p-value = 0.04202

Висновок на основі р-значення:
Відхиляємо H0: середній час від першого заходу на сайт до покупки (time-to-purchase) користувачів збільшився на рівні значущості 0.05
Висновок: редизайн сайту збільшив середній час від першого заходу на сайт до покупки (time-to-purchase) для користувачів

Висновок на основі t-статистичного та t-критичного:
Відхиляємо H0: середній час від першого заходу на сайт до покупки (time-to-purchase) користувачів збільшився на рівні значущості 0.05
Висновок: редизайн сайту збільшив середній час від першого заходу на сайт до покупки (time-to-purchase) для користувачів


При зміні рівня значущості на 0.05 рішення про відхилення гіпотези H0 не змінилося.

**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

In [3]:
import statsmodels.api as sm
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

count = 311   #кількість "успіхів"
nobs = 1200    #обсяг вибірки
p0 = 0.23      #гіпотетична пропорція (H0)

#z-тест для пропорції
stat, pval = proportions_ztest(count, nobs, p0, alternative='larger')
confint = proportion_confint(count, nobs, alpha=0.03, method='normal')

print("Тест для пропорції:")
print(f"Z-статистика = {stat:.3f}")
print(f"p-value = {pval:.4f}")
print(f"97% довірчий інтервал: {confint}")

Тест для пропорції:
Z-статистика = 2.306
p-value = 0.0106
97% довірчий інтервал: (0.23171700302179205, 0.28661633031154127)


1. Односторонній тест

   * $H_0: p = 0.23$
   * $H_1: p > 0.23$
   * Оскільки $p\text{-value} = 0.0106 < 0.03$, відхиляємо $H_0$.
     Оскільки є статистично значущі докази, що частка дорослих, які упізнають логотип компанії, **більша ніж 23%**.

2. 97% довірчий інтервал (0.2317, 0.2866)

   * Це означає, що з імовірністю 97% справжня частка лежить у межах **від 23.17% до 28.66%**.
   * Нижня межа за 0.23, тому в **двосторонньому тесті** не відхилили б $H_0$.
   * Але в **односторонньому тесті** — вже отримуємо значущість.